### ISSUE TITLE
BUG on splice ArraySchema

### ISSUE THREAD

**Issue by MateusMendesSantana** [colyseus/colyseus-unity3d#121]
Hi. When I remove items in my ArraySchema the OnAdd and OnRemove events are called on my Unity client. OnRemove executes correctly, but OnAdd (which should not even be called) returns empty values and ArraySchema now has incorrect values.
I modified the Decode function in the Schema.cs script and managed to solve this problem, could you take a look?
```
// Array type
else if (fieldType == "array")
{
  ISchemaCollection valueRef = (ISchemaCollection)(this[field] ?? Activator.CreateInstance(childType));
  ISchemaCollection currentValue = valueRef.Clone();

  int newLength = Convert.ToInt32(decode.DecodeNumber(bytes, it));
  int numChanges = Math.Min(Convert.ToInt32(decode.DecodeNumber(bytes, it)), newLength);

  bool hasRemoval = (currentValue.Count > newLength);
  hasChange = (numChanges > 0) || hasRemoval;

  bool hasIndexChange = false;

  // ensure current array has the same length as encoded one

  for (var i = 0; i < numChanges; i++)
  {
    var newIndex = Convert.ToInt32(decode.DecodeNumber(bytes, it));

    int indexChangedFrom = -1;
    if (decode.IndexChangeCheck(bytes, it))
    {
      decode.DecodeUint8(bytes, it);
      indexChangedFrom = Convert.ToInt32(decode.DecodeNumber(bytes, it));
      hasIndexChange = true;
    }

    var isNew = !hasIndexChange && !currentValue.ContainsKey(newIndex);

    if (currentValue.HasSchemaChild)
    {
      Schema item = null;

      if (isNew)
      {
        item = (Schema)CreateTypeInstance(bytes, it, currentValue.GetChildType());

      }
      else if (indexChangedFrom != -1)
      {
        item = (Schema)valueRef[indexChangedFrom];
      }
      else
      {
        item = (Schema)valueRef[newIndex];
      }

      if (item != null)
      {
        item.Decode(bytes, it);
        currentValue[newIndex] = item;
      }
    }
    else
    {
      currentValue[newIndex] = decode.DecodePrimitiveType(childPrimitiveType, bytes, it);
    }

    if (isNew)
    {
      currentValue.InvokeOnAdd(currentValue[newIndex], newIndex);
    }
    else
    {
      currentValue.InvokeOnChange(currentValue[newIndex], newIndex);
    }
  }

  if (hasRemoval)
  {
    IDictionary items = currentValue.GetItems();

    for (int i = newLength, l = currentValue.Count; i < l; i++)
    {
      var item = currentValue[i];
      if (item is Schema)
      {
        (item as Schema).OnRemove?.Invoke();
      }

      items.Remove(i);
      currentValue.InvokeOnRemove(item, i);
    }
  }
  value = currentValue;
}
```

**Comment by kevincastejon** [comment]
Same issue here, but the OnRemove is not pointing on the right index nor item.
On a 5 items array, if I remove the first one the OnRemove callback params will be the last item and the last index, then 4 OnAdd will be called back with empty item.
I've made a simple repo to reproduce this here :
https://github.com/kevincastejon/colyseus-unity3d-arrayschema-issue

**Comment by endel** [comment]
Fixed on 0.14 🎉  https://github.com/colyseus/colyseus/releases/tag/0.14.0
